From 55c1cdc490dd2fbc6d664deb2a9b27f0e7af3ec4 Mon Sep 17 00:00:00 2001
From: Nic Costa <nic.costa@arm.com>
Date: Tue, 24 Oct 2017 14:46:04 -0500
Subject: [PATCH 1/3] Adding FOTA progress

This patch adds support for re-using the display manager interface
with the workplace environmental monitor. This allows us to see
progress for checking the installation of the new firmware when
upgrading during boot.

This patch adds support for including the number of steps in the
printProgress message. The printProgress will prefix the message
with '[x/y] ', where 'x' is the current step number and 'y' is the
total number of steps.
The setStep function takes two arguments; the first is the current
step number and the second is the total number of steps.
If the total number of steps is 0 or the current step number is
greater than the total the prefixed message will not be printed.

Signed-off-by: Nic Costa <nic.costa@arm.com>
---
 source/active_application.cpp | 31 ++++++++++++++++++++++
 source/fota.cpp               | 60 +++++++++++++++++++++++++++++++++++++++++++
 source/fota.h                 | 34 ++++++++++++++++++++++++
 source/upgrade.cpp            |  3 +++
 4 files changed, 128 insertions(+)
 create mode 100644 source/fota.cpp
 create mode 100644 source/fota.h

diff --git a/source/active_application.cpp b/source/active_application.cpp
index 2f3be57..8b5c97f 100755
--- a/source/active_application.cpp
+++ b/source/active_application.cpp
@@ -20,6 +20,7 @@
 #define __STDC_FORMAT_MACROS
 #endif
 
+#include "fota.h"
 #include "active_application.h"
 #include "bootloader_common.h"
 
@@ -79,6 +80,11 @@ bool readActiveFirmwareHeader(arm_uc_firmware_details_t* details)
             if (event_callback == ARM_UC_PAAL_EVENT_GET_ACTIVE_FIRMWARE_DETAILS_DONE)
             {
                 result = true;
+
+                tr_info("Active firmware header");
+                tr_info("    version : %llu", details->version);
+                tr_info("    size    : %llu", details->size);
+                printSHA256(details->hash);
             }
         }
     }
@@ -151,6 +157,9 @@ int checkActiveApplication(arm_uc_firmware_details_t* details)
 #if defined(SHOW_PROGRESS_BAR) && SHOW_PROGRESS_BAR == 1
                 printProgress(details->size - remaining,
                               details->size);
+                fota::printProgress(details->size - remaining,
+                                    details->size,
+                                    "CHK Active");
 #endif
             }
 
@@ -189,6 +198,7 @@ bool eraseActiveFirmware(uint32_t firmwareSize)
     tr_debug("eraseActiveFirmware");
 
     /* get sector size of where the new firmware would end */
+    fota::printProgress(0, 100, "Erasing");
     uint32_t lastSectorSize =
         flash.get_sector_size(MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS +
                               MBED_CONF_APP_FIRMWARE_METADATA_HEADER_SIZE +
@@ -207,9 +217,12 @@ bool eraseActiveFirmware(uint32_t firmwareSize)
              (uint32_t) MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS + sizeRoundedUp);
 
     /* erase flash to make place for new application */
+    fota::printProgress(33, 100, "Erasing");
     int result = flash.erase(MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS,
                              sizeRoundedUp);
 
+    fota::printProgress(100, 100, "Erasing");
+
     return (result == 0);
 }
 
@@ -219,6 +232,7 @@ bool writeActiveFirmwareHeader(arm_uc_firmware_details_t* details)
 
     bool result = false;
 
+    fota::printProgress(0, 100, "Write HDR");
     if (details)
     {
         /* round up program size to nearest page size */
@@ -251,9 +265,11 @@ bool writeActiveFirmwareHeader(arm_uc_firmware_details_t* details)
             (output_buffer.size == ARM_UC_INTERNAL_HEADER_SIZE_V2))
         {
             /* write header using FlashIAP API */
+            fota::printProgress(33, 100, "Write HDR");
             int ret = flash.program(buffer_array,
                                     MBED_CONF_APP_FIRMWARE_METADATA_HEADER_ADDRESS,
                                     programSize);
+            fota::printProgress(100, 100, "Write HDR");
 
             result = (ret == 0);
         }
@@ -268,6 +284,7 @@ bool writeActiveFirmware(uint32_t index, arm_uc_firmware_details_t* details)
 
     bool result = false;
 
+    fota::printProgress(0, 100, "Write FW");
     if (details)
     {
         const uint32_t pageSize = flash.get_page_size();
@@ -338,6 +355,9 @@ bool writeActiveFirmware(uint32_t index, arm_uc_firmware_details_t* details)
 
 #if defined(SHOW_PROGRESS_BAR) && SHOW_PROGRESS_BAR == 1
                     printProgress(offset + programOffset, details->size);
+                    fota::printProgress(offset + programOffset,
+                                        details->size,
+                                        "Write FW");
 #endif
                 }
 
@@ -356,6 +376,8 @@ bool writeActiveFirmware(uint32_t index, arm_uc_firmware_details_t* details)
             }
         }
 
+        fota::printProgress(100, 100, "Write FW");
+
         result = (retval == 0);
     }
 
@@ -376,6 +398,7 @@ bool copyStoredApplication(uint32_t index,
     /* Step 1. Erase active application                                      */
     /*************************************************************************/
 
+    fota::setStep(2, 5);
     result = eraseActiveFirmware(details->size);
 
     /*************************************************************************/
@@ -384,6 +407,7 @@ bool copyStoredApplication(uint32_t index,
 
     if (result)
     {
+        fota::setStep(3, 5);
         result = writeActiveFirmwareHeader(details);
     }
 
@@ -393,6 +417,7 @@ bool copyStoredApplication(uint32_t index,
 
     if (result)
     {
+        fota::setStep(4, 5);
         result = writeActiveFirmware(index, details);
     }
 
@@ -404,10 +429,16 @@ bool copyStoredApplication(uint32_t index,
     {
         tr_info("Verify new active firmware:");
 
+        fota::setStep(5, 5);
         int recheck = checkActiveApplication(details);
 
         result = (recheck == RESULT_SUCCESS);
     }
 
+    if (!result) {
+        fota::setStep(0, 0);
+        fota::printProgress(0, 1, "UPDATE FAILED!");
+    }
+
     return result;
 }
diff --git a/source/fota.cpp b/source/fota.cpp
new file mode 100644
index 0000000..ca6cab2
--- /dev/null
+++ b/source/fota.cpp
@@ -0,0 +1,60 @@
+#include "fota.h"
+
+/* Since this structure needs to be accessed by both the setStep
+ * and printProgress calls we statically declare it here.
+ *
+ * @param current The step number in progress.
+ * @param total The number of total steps.
+ */
+static struct {
+    uint32_t current;
+    uint32_t total;
+} steps = { 0, 0 };
+
+/** Wrapper function for obtaining the active display manager. There is
+ * only a single occurring instance of this at any given time.
+ *
+ * @return Returns a pointer to the active display manager.
+ */
+DisplayMan *fota::getDisplayManager(void)
+{
+    static DisplayMan display;
+    display.init("FOTA");
+    display.set_downloading();
+    display.refresh();
+    return &display;
+}
+
+/** Prints the progress of the current activity to the LCD using the
+ * current active display manager.
+ *
+ * @param progress The current progress being made.
+ * @param total The total, or maximum, value of progress to be made.
+ * @param msg An optional message to be displayed (default '...')
+ */
+void fota::printProgress(uint32_t progress, uint32_t total, const char *msg)
+{
+    char buf[17] = {0}; /* assume 16 character LCD screen */
+
+    if (0 != steps.total && steps.current <= steps.total) {
+        snprintf(buf, sizeof(buf) - 1, "[%lu/%lu] %s", steps.current, steps.total, msg);
+    } else {
+        snprintf(buf, sizeof(buf) - 1, "%s", msg);
+    }
+
+    DisplayMan *display = getDisplayManager();
+    display->set_progress(buf, progress, total);
+    display->refresh();
+}
+
+/** Sets the current step number and the total number of steps to be prefixed
+ * to the printProgress message.
+ *
+ * @param current The step number in progress.
+ * @param total The number of total steps.
+ */
+void fota::setStep(uint32_t current, uint32_t total)
+{
+    steps.current = current;
+    steps.total = total;
+}
diff --git a/source/fota.h b/source/fota.h
new file mode 100644
index 0000000..5ad60b3
--- /dev/null
+++ b/source/fota.h
@@ -0,0 +1,34 @@
+#ifndef __BOOTLOADER_FOTA_H__
+#define __BOOTLOADER_FOTA_H__
+
+#include "displayman.h"
+
+namespace fota {
+
+/** Wrapper function for obtaining the active display manager. There is
+ * only a single occurring instance of this at any given time.
+ *
+ * @return Returns a pointer to the active display manager.
+ */
+DisplayMan *getDisplayManager(void);
+
+/** Prints the progress of the current activity to the LCD using the
+ * current active display manager.
+ *
+ * @param progress The current progress being made.
+ * @param total The total, or maximum, value of progress to be made.
+ * @param msg An optional message to be displayed (default '...')
+ */
+void printProgress(uint32_t progress, uint32_t total, const char *msg = "...");
+
+/** Sets the current step number and the total number of steps to be prefixed
+ * to the printProgress message.
+ *
+ * @param current The step number in progress.
+ * @param total The number of total steps.
+ */
+void setStep(uint32_t current, uint32_t total);
+
+} // end namespace fota
+
+#endif
diff --git a/source/upgrade.cpp b/source/upgrade.cpp
index bef539a..abf742f 100755
--- a/source/upgrade.cpp
+++ b/source/upgrade.cpp
@@ -20,6 +20,7 @@
 #define __STDC_FORMAT_MACROS
 #endif
 
+#include "fota.h"
 #include "upgrade.h"
 
 #include "update-client-paal/arm_uc_paal_update.h"
@@ -126,6 +127,7 @@ bool checkStoredApplication(uint32_t source,
 
 #if defined(SHOW_PROGRESS_BAR) && SHOW_PROGRESS_BAR == 1
             printProgress(offset, details->size);
+            fota::printProgress(offset, details->size, "CHK Flash");
 #endif
         }
 
@@ -328,6 +330,7 @@ bool upgradeApplicationFromStorage(void)
                         index);
 
                 /* Validate candidate firmware body. */
+                fota::setStep(1, 5);
                 bool firmwareValid = checkStoredApplication(index,
                                                             &imageDetails);
 
-- 
2.14.2

